### PR TITLE
OpenGL shader generation part 1

### PR DESCRIPTION
**Pull request by bunnei**
This is the first part of a set of changes to generate GLSL shaders for Switch vertex/fragment shaders. This adds basic shader decoding, an interface for managing shaders, and some code to generate simple shaders.

**THIS WILL NOT WORK WITH ANY GAMES YET**

**THIS CAN MOSTLY BE REVIEWED COMMIT BY COMMIT**
I'd like to get this merged soon. I'd suggest avoiding the large boilerplate changes from Citra, and focus on the few new modules, which are relatively small:
* gl_shader_manager
* gl_shader_decompiler
* shader_bytecode

### CODE REVIEW

**Comment by wwylele** [comment]
Seeing that my shader manager change to citra getting ported here, I just want to remind that the code here was an old version of mine, which contains some hacky code style. A new cleaned up version is at https://github.com/citra-emu/citra/pull/3645

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 1]
```diff
@@ -2,57 +2,158 @@
...
-                             const std::function<std::string(u32)>& inputreg_getter,
-                             const std::function<std::string(u32)>& outputreg_getter,
-                             bool sanitize_mul, const std::string& emit_cb,
-                             const std::string& setemit_cb) {
-    Impl impl(program_code, swizzle_data, main_offset, inputreg_getter, outputreg_getter,
-              sanitize_mul, emit_cb, setemit_cb);
-    return impl.Decompile();
+boost::optional<std::string> DecompileProgram(const ProgramCode& program_code, u32 main_offset) {
+    try {
+        auto subroutines = ControlFlowAnalyzer(program_code, main_offset).GetSubroutines();
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Should we add an ASSERT that the main method has AlwaysEnd as exit method? Just in case we find a shader that is too big and doesn't fit in the MAX_PROGRAM_CODE_LENGTH size.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 2]
```diff
@@ -280,18 +241,6 @@ void RasterizerOpenGL::DrawArrays() {
     // Sync and bind the texture surfaces
     BindTextures();
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: Should texture binding be moved to the inside of the shader program manager?

File: src/video_core/engines/maxwell_3d.h [hunk 3]
```diff
@@ -427,14 +427,11 @@ class Maxwell3D final {
                         BitField<0, 1, u32> enable;
                         BitField<4, 4, ShaderProgram> program;
                     };
-                    u32 start_id;
-                    INSERT_PADDING_WORDS(1);
-                    u32 gpr_alloc;
```
Comment (Subv) [inline src/video_core/engines/maxwell_3d.h]: Why did you remove gpr_alloc and type? please add them back.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 4]
```diff
@@ -130,7 +135,294 @@ class GLSLGenerator {
...
+    void GenerateDeclarations() {
+        for (const auto& reg : declr_register) {
+            declarations.AddLine("float " + reg + " = 0.0;");
+        }
+        declarations.AddLine("");
+
+        for (const auto& index : declr_input_attribute) {
+            // TODO(bunnei): Use proper number of elements for these
+            declarations.AddLine(
+                "layout(location = " + std::to_string(static_cast<u32>(index) - 8) + ") in vec4 " +
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: What is this 8?

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 5]
```diff
@@ -130,7 +135,294 @@ class GLSLGenerator {
...
+                GetInputAttribute(index) + ";");
+        }
+        declarations.AddLine("");
+
+        for (const auto& index : declr_output_attribute) {
+            // TODO(bunnei): Use proper number of elements for these
+            declarations.AddLine(
+                "layout(location = " + std::to_string(static_cast<u32>(index) - 8) + ") out vec4 " +
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: What is this 8?

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 6]
```diff
@@ -167,15 +171,69 @@ void RasterizerOpenGL::SetupVertexArray(u8* array_ptr, GLintptr buffer_offset) {
...
-    UNREACHABLE();
-}
+void RasterizerOpenGL::SetupShaders(u8* buffer_ptr, GLintptr buffer_offset, size_t ptr_pos) {
+    // Helper function for uploading uniform data
+    const auto copy_buffer = [&](GLuint handle, GLintptr offset, GLsizeiptr size) {
+        if (has_ARB_direct_state_access) {
+            glCopyNamedBufferSubData(stream_buffer->GetHandle(), handle, offset, 0, size);
+        } else {
+            glBindBuffer(GL_COPY_WRITE_BUFFER, handle);
+            glCopyBufferSubData(GL_ARRAY_BUFFER, GL_COPY_WRITE_BUFFER, offset, 0, size);
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: This seems wrong, why would you copy the vertex data into the uniform buffer?

File: src/video_core/renderer_opengl/gl_shader_manager.cpp [hunk 7]
```diff
@@ -38,6 +38,14 @@ void SetShaderSamplerBindings(GLuint shader) {
     cur_state.Apply();
 
     // Set the texture samplers to correspond to different texture units
+    for (u32 texture = 0; texture < 32; ++texture) {
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_manager.cpp]: Use an enum or the size of an existing array for this magic number

File: src/video_core/utils.h [hunk 8]
```diff
@@ -151,12 +151,15 @@ static inline void MortonCopyPixels128(u32 width, u32 height, u32 bytes_per_pixe
...
             u32 morton_offset =
                 GetMortonOffset128(x, y, bytes_per_pixel) + coarse_y * width * bytes_per_pixel;
-            u32 gl_pixel_index = (x + (height - 1 - y) * width) * gl_bytes_per_pixel;
+            u32 gl_pixel_index = (x + y * width) * gl_bytes_per_pixel;
 
             data_ptrs[morton_to_gl] = morton_data + morton_offset;
             data_ptrs[!morton_to_gl] = &gl_data[gl_pixel_index];
 
-            memcpy(data_ptrs[0], data_ptrs[1], bytes_per_pixel);
+            data_ptrs[0][0] = data_ptrs[1][3];
```
Comment (Subv) [inline src/video_core/utils.h]: This makes it so this function only works for bytes_per_pixel = 4. Is this intended?

File: src/video_core/engines/shader_bytecode.h [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        hex = instr.hex;
+        return *this;
+    }
+
+    OpCode opcode;
+    BitField<0, 8, Register> gpr1;
+    BitField<8, 8, Register> gpr2;
+    BitField<16, 4, Pred> pred;
+    BitField<39, 8, Register> gpr3;
+    BitField<45, 1, u64> nb;
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/video_core/engines/shader_bytecode.h]: Addressed in a later commit, this is negate_operand_b

File: src/video_core/engines/shader_bytecode.h [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+namespace Tegra {
+namespace Shader {
+
+enum class Pred : u64 {
+    UnusedIndex = 0x7,
+    NeverExecute = 0xf,
+};
+
+#pragma pack(1)
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/video_core/engines/shader_bytecode.h]: Addressed

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp (hunk 1 above)
Comment (bunnei) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: `ControlFlowAnalyzer` throws `DecompileFail` if it doesn't find it. This is probably as good as an `ASSERT` for now (shader decompilation will fail, which asserts)

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 2 above)
Comment (bunnei) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp]: Addressed in a later commit

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp (hunk 4 above)
Comment (bunnei) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Addressed

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp (hunk 5 above)
Comment (bunnei) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Addressed

File: src/video_core/utils.h (hunk 8 above)
Comment (bunnei) [reply src/video_core/utils.h]: Dropping this commit for now

File: src/video_core/engines/shader_bytecode.h [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+};
+static_assert(sizeof(OpCode) == 0x8, "Incorrect structure size");
+
+} // namespace Shader
+} // namespace Tegra
+
+namespace std {
+
+template <>
+struct make_unsigned<Tegra::Shader::Attribute> {
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/video_core/engines/shader_bytecode.h]: Added a TODO in code, and opened #330. I don't plan on fixing this with this PR unless someone proposes a simple solution, but we will fix it later.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 6 above)
Comment (bunnei) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp]: It's not, everything is uploaded in one stream buffer. This is how we are doing things [in Citra with glvtx](https://github.com/citra-emu/citra/pull/3499/files#diff-d1b91c65951bd9733c0e7d924161e57dR804), but I suspect this can change when we have your uniform support.

File: src/video_core/engines/shader_bytecode.h [hunk 12]
```diff
@@ -281,7 +282,6 @@ enum class SubOp : u64 {
     Rsq = 0x5,
 };
 
-#pragma pack(1)
```
Comment (Subv) [inline src/video_core/engines/shader_bytecode.h]: Remove the matching `pragma pack()` at the bottom

File: src/video_core/renderer_opengl/gl_shader_manager.cpp (hunk 7 above)
Comment (bunnei) [reply src/video_core/renderer_opengl/gl_shader_manager.cpp]: Addressed

File: src/video_core/engines/shader_bytecode.h (hunk 12 above)
Comment (bunnei) [reply src/video_core/engines/shader_bytecode.h]: Addressed
